### PR TITLE
[NFC] Refactor parameter counting

### PR DESCRIPTION
**Pull request by CodaFi**
As a first test of the new function representation, use it to count parameters in CSGen.

Depends on #10264

Pre-rebase the relevant code is in `CSGen.cpp`

### CODE REVIEW

File: include/swift/AST/Types.h [hunk 1]
```diff
@@ -2282,7 +2282,35 @@ getSILFunctionLanguage(SILFunctionTypeRepresentation rep) {
...
+struct CallArgParam {
+  /// The type of the argument or parameter. For a variadic parameter,
+  /// this is the element type.
+  Type Ty;
+  
+  // The label associated with the argument or parameter, if any.
+  Identifier Label;
+  
+  /// Whether the parameter has a default argument.  Not valid for arguments.
+  bool HasDefaultArgument = false;
```
Comment (DougGregor) [inline include/swift/AST/Types.h]: `HasDefaultArgument` doesn't belong in the type. I think you will need to use something other than `CallArgParam` for your storage.

File: include/swift/AST/Types.h [hunk 2]
```diff
@@ -2295,7 +2323,8 @@ getSILFunctionLanguage(SILFunctionTypeRepresentation rep) {
 /// be 'thin', indicating that a function value has no capture context and can be
 /// represented at the binary level as a single function pointer.
 class AnyFunctionType : public TypeBase {
-  const Type Input;
+  const ArrayRef<CallArgParam> Input;
```
Comment (DougGregor) [inline include/swift/AST/Types.h]: I'd prefer to tail-allocate the parameters in the subclasses, so we don't pay for this extra pointer in such a common type.

File: lib/AST/ASTContext.cpp [hunk 3]
```diff
@@ -3136,6 +3136,16 @@ getGenericFunctionRecursiveProperties(Type Input, Type Result) {
   return properties;
 }
 
+AnyFunctionType::AnyFunctionType(TypeKind Kind, const ASTContext *CanTypeContext,
+                ArrayRef<CallArgParam> Input, Type RawInput, Type Output,
+                RecursiveTypeProperties properties,
+                const ExtInfo &Info)
+: TypeBase(Kind, CanTypeContext, properties),
+  Input((CanTypeContext ?: &RawInput->getASTContext())->AllocateCopy(Input)),
```
Comment (DougGregor) [inline lib/AST/ASTContext.cpp]: Hmm... I'm confused why we're doing the `CanTypeContext` null check dance here. If we do the tail allocation thing, then don't need to do the AllocateCopy dance here... we just copy over the data we get.

**Comment by CodaFi** [comment]
I destroyed any hope of a reasonable merge here.  Going to reopen this elsewhere.

**Pull request closed without merging**